jina/clients/__init__.py

```diff
@@ -1,91 +1,169 @@
+            raise BadInputFunction from ex
+
+    def get_requests(self, **kwargs) -> Iterator['Request']:
+        """Get request in generator"""
+        _kwargs = vars(self.args)
+        _kwargs['data'] = self.input_fn
+        # override by the caller-specific kwargs
+        _kwargs.update(kwargs)
+
+        tname = str(self.mode).lower()
+        if 'mode' in kwargs:
+            tname = str(kwargs['mode']).lower()
+
+        return getattr(request, tname)(**_kwargs), tname
+
+    @property
+    def input_fn(self) -> InputFnType:
+        """ An iterator of bytes, each element represents a document's raw content,
+        i.e. ``input_fn`` defined in the protobuf
+        """
+        if self._input_fn is not None:
+            return self._input_fn
+        else:
+            raise BadClient('input_fn is empty or not set')
+
+    @input_fn.setter
+    def input_fn(self, bytes_gen: InputFnType) -> None:
+        if hasattr(bytes_gen, '__call__'):
+            self._input_fn = bytes_gen()
+        else:
+            self._input_fn = bytes_gen
+
+    def train(self, input_fn: Optional[InputFnType] = None,
+              output_fn: Callable[['Request'], None] = None, **kwargs) -> None:
+        self.mode = RequestType.TRAIN
+        self._submit(input_fn, output_fn, **kwargs)
+
+    def search(self, input_fn: Optional[InputFnType] = None,
+               output_fn: Callable[['Request'], None] = None, **kwargs) -> None:
+        self.mode = RequestType.SEARCH
+        self._submit(input_fn, output_fn, **kwargs)
+
+    def index(self, input_fn: Optional[InputFnType] = None,
+              output_fn: Callable[['Request'], None] = None, **kwargs) -> None:
+        self.mode = RequestType.INDEX
+        self._submit(input_fn, output_fn, **kwargs)
+
+    def _submit(self,
+                input_fn: Callable,
+                on_done: Callable,
+                on_error: Callable = None,
+                on_always: Callable = None, **kwargs):
+
+        async def _inner():
             try:
-                msg = zmqlet.sock.recv()
-                if msg == b'TERMINATE':
-                    # ideal way of exit, but PyClient socket might have closed before we recv it here
-                    logger.debug('received terminate message from the client response stream')
-                    zmqlet.sock.send_string('')
-                    break
+                self.input_fn = input_fn
+                req_iter, tname = self.get_requests(**kwargs)
+                async with grpc.aio.insecure_channel(f'{self.args.host}:{self.args.port_expose}',
+                                                     options=[('grpc.max_send_message_length', -1),
+                                                              ('grpc.max_receive_message_length', -1)]) as channel:
+                    stub = jina_pb2_grpc.JinaRPCStub(channel)
+                    self.logger.success(f'connected to the gateway at {self.args.host}:{self.args.port_expose}!')
```

Comment:
`await channel.ready()` here before the success message

Reply:
https://github.com/grpc/grpc/blob/master/examples/python/helloworld/async_greeter_client.py